File: src-tauri/src/domain/source.rs
use serde::Serialize;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SourceAccount {
    pub provider: String,
    pub provider_account_id: String,
    pub display_name: Option<String>,
    pub client_id: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SourceCollection {
    pub source_account_id: i64,
    pub provider_collection_id: String,
    pub kind: String,
    pub name: String,
    pub snapshot_id: Option<String>,
    pub owner_provider_id: Option<String>,
    pub is_accessible: bool,
    pub access_issue: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SourceTrack {
    pub provider: String,
    pub provider_track_id: String,
    pub uri: Option<String>,
    pub isrc: Option<String>,
    pub title: String,
    pub normalized_title: String,
    pub artists_json: String,
    pub normalized_artists: String,
    pub album: Option<String>,
    pub normalized_album: Option<String>,
    pub duration_ms: Option<i64>,
    pub disc_number: Option<i64>,
    pub track_number: Option<i64>,
    pub release_year: Option<i64>,
    pub explicit: Option<bool>,
    pub version_kind: Option<String>,
    pub version_detail: Option<String>,
    pub image_url: Option<String>,
    pub external_url: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CollectionEntry {
    pub position: i64,
    pub source_track_id: Option<i64>,
    pub provider_item_uri: Option<String>,
    pub item_type: String,
    pub added_at: Option<i64>,
    pub unavailable_reason: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SourceCollectionItem {
    pub position: i64,
    pub track: Option<SourceTrack>,
    pub provider_item_uri: Option<String>,
    pub item_type: String,
    pub added_at: Option<i64>,
    pub unavailable_reason: Option<String>,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct SourceAccountOverview {
    pub display_name: Option<String>,
    pub last_source_sync_at: Option<i64>,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct SourceCollectionSummary {
    pub id: i64,
    pub provider_collection_id: String,
    pub kind: String,
    pub name: String,
    pub is_accessible: bool,
    pub access_issue: Option<String>,
    pub entry_count: i64,
    pub tracked_by_default: bool,
    pub tracked_entry_count: i64,
    pub image_url: Option<String>,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct SpotifySourceOverview {
    pub account: Option<SourceAccountOverview>,
    pub liked_songs: Option<SourceCollectionSummary>,
    pub playlist_count: i64,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct SourceCollectionListPage {
    pub items: Vec<SourceCollectionSummary>,
    pub total: i64,
    pub offset: u32,
    pub limit: u32,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct SourceTrackView {
    pub id: i64,
    pub provider_track_id: String,
    pub title: String,
    pub artists: Vec<String>,
    pub album: Option<String>,
    pub release_year: Option<i64>,
    pub duration_ms: Option<i64>,
    pub explicit: Option<bool>,
    pub image_url: Option<String>,
    pub external_url: Option<String>,
    pub local_present: bool,
    pub local_format: Option<String>,
    pub match_state: String,
    pub acquisition_status: Option<String>,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct SourceCollectionEntryView {
    pub position: i64,
    pub item_type: String,
    pub added_at: Option<i64>,
    pub unavailable_reason: Option<String>,
    pub track: Option<SourceTrackView>,
    pub tracking_included: bool,
    pub tracking_overridden: bool,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct SourceCollectionPage {
    pub collection: SourceCollectionSummary,
    pub entries: Vec<SourceCollectionEntryView>,
    pub total: i64,
    pub offset: u32,
    pub limit: u32,
}

File: src/lib/source.ts
import { invoke } from '@tauri-apps/api/core';
import type { InvokeFn } from './app-info';

export interface SourceAccountOverview {
  displayName: string | null;
  imageUrl: string | null;
  lastSourceSyncAt: number | null;
}

export interface SourceAlbumMetadata {
  artists: string[];
  releaseDate: string | null;
  albumType: string | null;
  label: string | null;
  copyrights: string[];
  externalUrl: string | null;
}

export interface SourceCollectionSummary {
  id: number;
  providerCollectionId: string;
  kind: string;
  name: string;
  isAccessible: boolean;
  accessIssue: string | null;
  entryCount: number;
  trackedByDefault: boolean;
  trackedEntryCount: number;
  localEntryCount: number;
  attentionEntryCount: number;
  imageUrl: string | null;
  externalUrl?: string | null;
  albumMetadata?: SourceAlbumMetadata | null;
}

export interface SpotifySourceOverview {
  account: SourceAccountOverview | null;
  likedSongs: SourceCollectionSummary | null;
  playlistCount: number;
}

export interface SourceCollectionListPage {
  items: SourceCollectionSummary[];
  total: number;
  offset: number;
  limit: number;
}

export interface SourceTrackView {
  id: number;
  providerTrackId: string;
  title: string;
  artists: string[];
  album: string | null;
  releaseYear: number | null;
  durationMs: number | null;
  explicit: boolean | null;
  imageUrl: string | null;
  externalUrl: string | null;
  localPresent: boolean;
  localFormat: string | null;
  matchState: 'matched' | 'unmatched' | string;
  acquisitionStatus: string | null;
}

export interface SourceCollectionEntryView {
  position: number;
  itemType: string;
  addedAt: number | null;
  unavailableReason: string | null;
  track: SourceTrackView | null;
  trackingIncluded: boolean;
  trackingOverridden: boolean;
}

export interface SourceCollectionPage {
  collection: SourceCollectionSummary;
  entries: SourceCollectionEntryView[];
  total: number;
  offset: number;
  limit: number;
}

export interface SpotifySourceHydration {
  overview: SpotifySourceOverview;
  playlists: SourceCollectionListPage;
  savedAlbums: SourceCollectionListPage;
}

export function getSpotifySourceOverview(
  invokeFn: InvokeFn = invoke,
): Promise<SpotifySourceOverview> {
  return invokeFn<SpotifySourceOverview>('get_spotify_source_overview');
}

export function listSpotifyPlaylists(
  offset = 0,
  limit = 100,
  invokeFn: InvokeFn = invoke,
): Promise<SourceCollectionListPage> {
  return invokeFn<SourceCollectionListPage>('list_spotify_playlists', {
    offset,
    limit,
  });
}

export function listSpotifySavedAlbums(
  offset = 0,
  limit = 100,
  invokeFn: InvokeFn = invoke,
): Promise<SourceCollectionListPage> {
  return invokeFn<SourceCollectionListPage>('list_spotify_saved_albums', {
    offset,
    limit,
  });
}

export function getSourceCollectionPage(
  collectionId: number,
  offset = 0,
  limit = 200,
  invokeFn: InvokeFn = invoke,
): Promise<SourceCollectionPage> {
  return invokeFn<SourceCollectionPage>('get_source_collection_page', {
    collectionId,
    offset,
    limit,
  });
}

export function setSourceCollectionTracking(
  collectionId: number,
  included: boolean,
  invokeFn: InvokeFn = invoke,
): Promise<void> {
  return invokeFn<void>('set_source_collection_tracking', {
    collectionId,
    included,
  });
}

export function setSourceTrackTracking(
  collectionId: number,
  sourceTrackId: number,
  included: boolean | null,
  invokeFn: InvokeFn = invoke,
): Promise<void> {
  return invokeFn<void>('set_source_track_tracking', {
    collectionId,
    sourceTrackId,
    included,
  });
}

export function setSourceTracksTracking(
  collectionId: number,
  sourceTrackIds: number[],
  included: boolean | null,
  invokeFn: InvokeFn = invoke,
): Promise<void> {
  return invokeFn<void>('set_source_tracks_tracking', {
    collectionId,
    sourceTrackIds,
    included,
  });
}

export async function hydrateSpotifySource(
  invokeFn: InvokeFn = invoke,
): Promise<SpotifySourceHydration> {
  const [overview, playlists, savedAlbums] = await Promise.all([
    getSpotifySourceOverview(invokeFn),
    listSpotifyPlaylists(0, 100, invokeFn),
    listSpotifySavedAlbums(0, 100, invokeFn),
  ]);
  return { overview, playlists, savedAlbums };
}

export function formatTrackDuration(durationMs: number | null): string {
  if (durationMs === null || durationMs < 0) {
    return '—';
  }

  const totalSeconds = Math.floor(durationMs / 1000);
  const minutes = Math.floor(totalSeconds / 60);
  const seconds = totalSeconds % 60;
  return `${minutes}:${seconds.toString().padStart(2, '0')}`;
}
